x-pack/filebeat/input/awss3/input.go

```diff
@@ -300,20 +304,26 @@ func (in *s3Input) createS3Lister(ctx v2.Context, cancelCtx context.Context, cli
 	return s3Poller, nil
 }
 
-func getRegionFromQueueURL(queueURL string, endpoint string) (string, error) {
+var errBadQueueURL = errors.New("QueueURL is not in format: https://sqs.{REGION_ENDPOINT}.{ENDPOINT}/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+
+func getRegionFromQueueURL(queueURL string, endpoint, defaultRegion string) (string, error) {
```

Comment:
We have both `regionName` from queue_url and `in.config.RegionName` in this function. Can we not perform all logical operations in this method and return a regionName back. It could be configured / from url / err

If err , then just `return fmt.Errorf("failed to get AWS region from queue_url: %w", err)`

The idea is that all the logic for retrieving the regionName shall be at one place. WDYT?

```diff
diff --git a/x-pack/filebeat/input/awss3/input.go b/x-pack/filebeat/input/awss3/input.go
index 4f45f91414..2379de2765 100644
--- a/x-pack/filebeat/input/awss3/input.go
+++ b/x-pack/filebeat/input/awss3/input.go
@@ -115,15 +115,11 @@ func (in *s3Input) Run(inputContext v2.Context, pipeline beat.Pipeline) error {
        defer cancelInputCtx()
 
        if in.config.QueueURL != "" {
-               regionName, err := getRegionFromQueueURL(in.config.QueueURL, in.config.AWSConfig.Endpoint, in.config.RegionName)
-               if err != nil && in.config.RegionName == "" {
+               regionName, err := getRegionFromQueueURL(inputContext, in.config.QueueURL, in.config.AWSConfig.Endpoint, in.config.RegionName)
+               if err != nil {
                        return fmt.Errorf("failed to get AWS region from queue_url: %w", err)
                }
-               if regionName != in.config.RegionName {
-                       inputContext.Logger.Warnf("configured region disagrees with queue_url region: %q != %q: using %[1]q",
-                               in.config.RegionName, regionName)
-                       regionName = in.config.RegionName
-               }
+
                in.awsConfig.Region = regionName
 
                // Create SQS receiver and S3 notification processor.
@@ -306,21 +302,29 @@ func (in *s3Input) createS3Lister(ctx v2.Context, cancelCtx context.Context, cli
 
 var errBadQueueURL = errors.New("QueueURL is not in format: https://sqs.{REGION_ENDPOINT}.{ENDPOINT}/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
 
-func getRegionFromQueueURL(queueURL string, endpoint, defaultRegion string) (string, error) {
-       // get region from queueURL
+// get region from the queue_url. Returns a configured region as a default
+// and error if the region cannot be parsed from the url and a default configuration is not set
+func getRegionFromQueueURL(ctx v2.Context, queueURL string, endpoint, defaultRegion string) (string, error) {
        // Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
        u, err := url.Parse(queueURL)
        if err != nil {
                return "", fmt.Errorf(queueURL + " is not a valid URL")
        }
+       var regionName string
        if (u.Scheme == "https" || u.Scheme == "http") && u.Host != "" {
                queueHostSplit := strings.SplitN(u.Host, ".", 3)
                if len(queueHostSplit) == 3 {
                        if queueHostSplit[2] == endpoint || (endpoint == "" && strings.HasPrefix(queueHostSplit[2], "amazonaws.")) {
-                               return queueHostSplit[1], nil
+                               regionName = queueHostSplit[1]
+
+                               if defaultRegion != "" && regionName != defaultRegion {
+                                       ctx.Logger.Warnf("configured region disagrees with queue_url region: %q != %q: using %[1]q",
+                                               defaultRegion, regionName)
+                                       regionName = defaultRegion
+                               }
                        }
                } else if defaultRegion != "" {
-                       return defaultRegion, nil
+                       regionName = defaultRegion
                }
+               return regionName, nil
        }
        return "", errBadQueueURL
diff --git a/x-pack/filebeat/input/awss3/input_test.go b/x-pack/filebeat/input/awss3/input_test.go
index 02a91022f5..30642c415b 100644
--- a/x-pack/filebeat/input/awss3/input_test.go
+++ b/x-pack/filebeat/input/awss3/input_test.go
@@ -8,6 +8,7 @@ import (
        "errors"
        "testing"
 
+       v2 "github.com/elastic/beats/v7/filebeat/input/v2"
        "github.com/stretchr/testify/assert"
 )
 
@@ -52,6 +53,7 @@ func TestGetProviderFromDomain(t *testing.T) {
 
 func TestGetRegionFromQueueURL(t *testing.T) {
        tests := []struct {
+               ctx      v2.Context
                name     string
                queueURL string
                endpoint string
@@ -102,7 +104,7 @@ func TestGetRegionFromQueueURL(t *testing.T) {
 
        for _, test := range tests {
                t.Run(test.name, func(t *testing.T) {
-                       got, err := getRegionFromQueueURL(test.queueURL, test.endpoint, test.deflt)
+                       got, err := getRegionFromQueueURL(test.ctx, test.queueURL, test.endpoint, test.deflt)
                        if !sameError(err, test.wantErr) {
                                t.Errorf("unexpected error: got:%v want:%v", err, test.wantErr)
                        }
```

Reply:
I'd rather not. Moving the logging into what is essentially a helper convolutes the code.

Reply:
The logging part I agree , yeah. Nvm, Can leave this as is.